common/logging/EventAuditLogger.php

```diff
@@ -0,0 +1,1080 @@
+            $msg = $this->create_rfc3881_msg($user, $group, $event, $patient_id, $outcome, $comments);
+            fwrite($conn, $msg);
+            fclose($conn);
+        }
+    }
+
+    /**
+     * Add an entry into the audit log table, indicating that an
+     * SQL query was performed. $outcome is true if the statement
+     * successfully completed.  Determine the event type based on
+     * the tables present in the SQL query.
+     *
+     * @param $statement
+     * @param $outcome
+     * @param null      $binds
+     */
+    public function auditSQLEvent($statement, $outcome, $binds = null)
+    {
+        $user =  $_SESSION['authUser'] ?? "";
+
+        /* Don't log anything if the audit logging is not enabled. Exception for "emergency" users */
+        if (!isset($GLOBALS['enable_auditlog']) || !($GLOBALS['enable_auditlog'])) {
+            if ((soundex($user) != soundex("emergency")) && (soundex($user) != soundex("breakglass"))) {
+                return;
+            }
+        }
+
+        $statement = trim($statement);
+
+        /**
+         * Don't audit SQL statements done to the audit log,
+         * or we'll have an infinite loop.
+         * Skip SELECT count() statements.
+         * Skip the SELECT made by the authCheckSession() function.
+         */
+        if ((stripos($statement, "insert into log") !== false)  // avoid infinite loop
+            || (stripos($statement, "FROM log ") !== false)        // avoid infinite loop
+            || (stripos($statement, "SELECT count(") === 0)        // Skip SELECT count() statements.
+            || (stripos($statement, "select username, password from users") === 0) // Skip the SELECT made by the authCheckSession() function.
+        ) {
+            return;
+        }
+
+        $group = isset($_SESSION['authGroup']) ?  $_SESSION['authGroup'] : "";
+        $comments = $statement;
+        $success = (int)($outcome !== false);
+        $checksum = ($outcome !== false) ? $this->sql_checksum_of_modified_row($statement) : '';
+
+        if (is_array($binds) && !empty($processed_binds)) {
+            $processed_binds = "";
+            // Need to include the binded variable elements in the logging
+            foreach ($binds as $value_bind) {
+                $processed_binds .= "'" . add_escape_custom($value_bind) . "',";
+            }
+            rtrim($processed_binds, ',');
+            $comments .= " (" . $processed_binds . ")";
+        }
+
+        // ViSolve : Don't log sequences - to avoid the affect due to GenID calls
+        if (strpos($comments, "sequences") !== false) {
+            return;
+        }
+
+        /* Determine the query type (select, update, insert, delete) */
+        $querytype = "select";
+        $querytypes = array("select", "update", "insert", "delete","replace");
+        foreach ($querytypes as $qtype) {
+            if (stripos($statement, $qtype) === 0) {
+                $querytype = $qtype;
+            }
+        }
```

Comment:
Also, it would seem like that this would be the most efficient place to then see if need to go ahead or if can exit dependent on the global 'Audit Logging SELECT Query' in adminsitration->globals->logging (ie. if this is is a select, select logging is turned off, and user is not a breakglass user, then get the heck out.